### ISSUE TITLE
Feature request: Support indexes on relationships

### ISSUE THREAD

**Issue by LuanaLoubet** [neo4j/neo4j#7225]
I need to create indexes in relationship to optimize my query. Cypher not support indexes in relationship ?
And I cannot create the follow constraint in relationship SHARED_REACTION:
CREATE CONSTRAINT ON ()-[sr:SHARED_REACTION]-() ASSERT EXISTS (sr.id)
Neo4j returns: 
"Unable to create CONSTRAINT.."
Neo.DatabaseError.Schema.ConstraintCreationFailed

How can I create this constraint and to use indexes in relationship SHARED_REACTION ?

**Comment by jexp** [comment]
You can create a property existence constraint in Neo4j Enterprise.

Why do you need to look up relationships by id ? What is your graph model & use-case?

**Comment by LuanaLoubet** [comment]
I am trying to create a index in a relationship in order to make the following matching process faster:

MATCH (s1:RxNorm)-->(r:Reaction)<--(s2:RxNorm) WHERE s1.id > s2.id
MERGE (s1)-[sd:SHARED_REACTION]-(s2)
     ON CREATE SET sd.n = 1
     ON MATCH SET sd.n = sd.n + 1

In this case, :RxNorm are drugs and :Reaction are reactions to the drugs. Each edge :SHARED_REACTION will count how many reactions are shared among a given pairs of drugs.

**Comment by spacecowboy** [comment]
One option which would enable you to use indexes would be to use an intermediate node, e.g.

```
(s1)-[:SHARES]->(:SHARED_REACTION)<-[:SHARES]-(s2)
```

Now, a constraint/index is possible:

```
CREATE CONSTRAINT ON (sr:SHARED_REACTION) ASSERT EXISTS (sr.id)
```

**Comment by rationalism** [comment]
I'd also like to request support for indexes on relationships. This isn't that important, because as @spacecowboy said, you can just use an intermediate node. However, putting indexes on relationships would make the graph schema simpler. Eg., if we have:

(n:email {subject: "Hey Bob"})-[:mentions {id: "xxxxxx"}]->(p:person {name: "Bob"})

there are two types of nodes, and one type of relationship. However, with an intermediate node, you'd get something like:

(n:email {subject: "Hey Bob"})-[:mentions]->(m:mention {id: "xxxxxx"}])-[:refers]->(p:person {name: "Bob"})

which is three types of nodes and two types of relationships. It does work, it's just a bit more complex.

**Comment by ehx-v1** [comment]
@rationalism you mean a PRETTY bit more complex. Imagine this for, let's say, 100 `mentions` relations, or already 25, for 25 the upper version would even still be human-readable, while the lower version would not.
Generally, intermediate nodes are less readable. However, in the given use case, it would actually even be a more senseful modelling, as shared reactions are rather another object than some abstract interconnection between objects (but of course both views are discussible). To back up this statement, I'd like to mention that it's described by substantive, which is typical of objects/nodes, and not by verb, which would fit a relationship.
@LuanaLoubet you are doing medical analysis on drugs with Neo4j? Interesting how different use cases Neo4j is used for! I'm mainly working on routing over the Spatial extension, with offpoints and stuff like that.
@spacecowboy I'd rather categorize this an enhancement request.

**Comment by spacecowboy** [comment]
Changed the labels and the title to match

**Comment by 0xfff** [comment]
+1
Because we constantly add relationships from different data sources (and therefore there could be duplicates) i would like to MERGE (vs create) a relationship.
I reason that a UNIQUE CONSTRAINT would help us there

**Comment by 0xfff** [comment]
I just tried MERGE and it works correctly (ie no duplicate edges were created)
is there only one relationship (of a given type) allowed between nodes ?

**Comment by spacecowboy** [comment]
@0xfff no you can have any number of relationships you want. But MERGE will not create a new relationship if one that satisfies the pattern already exists. So for your usecase MERGE is likely sufficient.

**Comment by MobiusF** [comment]
Can we vote for this feature??
I work with the spatial plugin and only want to have the geometry of a relationship (real world utility connection) in the spatial index. Now I have to have millions of bogus nodes.
DO NOT POLLUTE THE DATAMODEL!!
What's the status for this?? Will it ever happen??

**Comment by MerrickClark** [comment]
I'm assuming this is a non-trivial change considering the kernel tag. If at all possible, would be a great feature for an upcoming major release. Love Neo4j, you guys have another convert ;-)

**Comment by Lojjs** [comment]
Thanks for your feature request! Currently we have a long backlog of things we want to do in the Cypher Team, but I will add your request so it can be prioritized against other potential features.

**Comment by computationdoc** [comment]
I too am looking for indexing on relationship properties.

**Comment by Artgit** [comment]
+1 for this feature. Do you have any idea when it can be implemented in Neo4j?

**Comment by raymond1985** [comment]
+1 waiting for this feature, hope it will be implemented soon

**Comment by ctytgat** [comment]
+1

**Comment by Joe2001** [comment]
+1

Graph DBs promote, that the real data is the structure, which should be modeled by the links/relationships. Relationships should be first-class-citizens! So I cannot understand, that indexes on relationships are currently not supported!

See this simply case:

Let's have sessions and products (or other target objects), linked by "actions". I want to query the most active products by the the time stored as property in relationship :action:

match (s:Session)-[l:action]->(n:Product)
where l.time >= 1526511600.0
and l.time <= 1527116400.0
with distinct n,
count(l) as num_of_actions
return n.key, num_of_actions
order by num_of_actions desc
limit 100

Takes 10 minutes on our graph (> 10 millions of actions)!

You may answer: actions should be nodes... put all that stuff as/into nodes. But if I always have to follow that rule, I can step back to SQL DBs. And using intermediate nodes??? If I need nodes to make my relationships work...? Funny! Sounds like the good old 'cross table' on my SQL DB ;-)

**Comment by nhmelo** [comment]
+1

Currently working on a project strongly dependent on Neo4j and adding intermediate nodes would pollute the model immensely. I really don't like that solution and it seems like a needless workaround to actually having indexes on relationships.

Hopefully this feature becomes available soon! Thanks

**Comment by calvinalvin** [comment]
+1

**Comment by chrisvest** [comment]
There are at least three parts to this: First is being able to index relationships. Second is being able to apply a constraint on the indexed relationships. And third extending the Cypher planner smarts to take relationships indexes and constraints into account for planning. The first part, indexing relationships, is being worked on to some extent, but it will still be a while before the support has been added throughout the stack, to enable the functionality that is expected by this issue.

**Comment by go-spider** [comment]
looking forward,it's really slow to query relationship.

**Comment by ragavendraselvam-fourkites** [comment]
+1

**Comment by ujfjhz** [comment]
+1

**Comment by Icrush** [comment]
+1 for index on relationship

**Comment by chrisvest** [comment]
In 3.5, the new fulltext schema indexes can index relationships. So some of this functionality is now available. Support in normal schema indexes and constraints will come later.

**Comment by MobiusF** [comment]
@chrisvest : to add to the stack: for spatial indexes ( currently spatial is worked on to add it to the core!!! , I mean actual spatial, not just fiddling with points ), we need the (rtree) indexes of nodes and relationships to JOIN.

**Comment by chrisvest** [comment]
@MobiusF that's a separate feature, not related to this issue.

**Comment by makpcheng** [comment]
+1 for index on relationships

**Comment by cissecedric** [comment]
Hi, 
Just to get some news about this feature "Support indexes on relationships"..
I'm new to neo4j, I'm wandering, has it be done/is it planned ? 
It seems (not sure...) that query that involves relationships are not the fastest, it should be fatser if we could define indexes on relationships ?

**Comment by InverseFalcon** [comment]
@cissecedric It's implemented, here's the [docs for usage](https://neo4j.com/docs/cypher-manual/current/schema/index/#schema-index-fulltext-search). Note that this uses procedures to create and query the index, so it isn't quite as easy to use as the regular schema indexes.

That said, it does allow case insensitive lookup, allows indexing across nodes of multiple labels and properties (for the fulltext node indexes), or across multiple relationship types and properties (for the fulltext relationship indexes), and you can use different lucene analyzers when you create indexes this way.

As for whether it will help with your queries, keep in mind that the primary usage of indexes in Neo4j is for finding the starting points in your graph, after which we use traversals and filtering to find the rest of the patten (Neo4j uses index-free adjacency over table joins). So if you need to find starting relationships in your graph by their type(s) and property (or properties) then this may help you quite a bit.

But if you are already using index lookups on your nodes to quickly find the starting points in your graph, and you just need to traverse and filter, then this likely won't be helpful, as you may already be working with a small and relevant set of connected relationships at that point already, and so a relationship index lookup may not help at all (returning far more results needing to be filtered than the set of relationships you already have).

**Comment by charleskoehl** [comment]
> @cissecedric It's implemented, here's the [docs for usage](https://neo4j.com/docs/cypher-manual/current/schema/index/#schema-index-fulltext-search). Note that this uses procedures to create and query the index, so it isn't quite as easy to use as the regular schema indexes.

That links to the docs for full-text searching.

**Comment by InverseFalcon** [comment]
@charleskoehl 
It does. The fulltext index can be created on relationships as well as nodes, using the procedures `db.index.fulltext.createRelationshipIndex()` and `db.index.fulltext.queryRelationships()`. Currently this supports indexes on string values across multiple relationship types and multiple relationship properties.

**Comment by ismetaydin** [comment]
+1

**Comment by selvarajendran** [comment]
> @charleskoehl
> It does. The fulltext index can be created on relationships as well as nodes, using the procedures `db.index.fulltext.createRelationshipIndex()` and `db.index.fulltext.queryRelationships()`. Currently this supports indexes on string values across multiple relationship types and multiple relationship properties.

Could you please let me know, where I can check the example of db.index.fulltext.createRelationshipIndex() and db.index.fulltext.queryRelationships(). I am looking for one. Thanks in advance.

**Comment by burqen** [comment]
> > @charleskoehl
> > It does. The fulltext index can be created on relationships as well as nodes, using the procedures `db.index.fulltext.createRelationshipIndex()` and `db.index.fulltext.queryRelationships()`. Currently this supports indexes on string values across multiple relationship types and multiple relationship properties.
> 
> Could you please let me know, where I can check the example of db.index.fulltext.createRelationshipIndex() and db.index.fulltext.queryRelationships(). I am looking for one. Thanks in advance.

Here is full-text indexes described in documentation :) 
https://neo4j.com/developer/kb/fulltext-search-in-neo4j/
https://neo4j.com/docs/cypher-manual/current/administration/indexes-for-full-text-search/

**Comment by selvarajendran** [comment]
> > > @charleskoehl
> > > It does. The fulltext index can be created on relationships as well as nodes, using the procedures `db.index.fulltext.createRelationshipIndex()` and `db.index.fulltext.queryRelationships()`. Currently this supports indexes on string values across multiple relationship types and multiple relationship properties.
> > 
> > 
> > Could you please let me know, where I can check the example of db.index.fulltext.createRelationshipIndex() and db.index.fulltext.queryRelationships(). I am looking for one. Thanks in advance.
> 
> Here is full-text indexes described in documentation :)
> https://neo4j.com/developer/kb/fulltext-search-in-neo4j/
> https://neo4j.com/docs/cypher-manual/current/administration/indexes-for-full-text-search/

Thank you.

**Comment by selvarajendran** [comment]
@burqen

I tried below with full text index as below

CALL db.index.fulltext.queryNodes("index1", "x") YIELD node as node1, score as score1
With collect({id: node1.id, score: score1}) as rows1
CALL db.index.fulltext.queryNodes("index2", "name:Y") YIELD node as node2, score as score2
With collect({id: node2.id, score: score2}) as rows2, rows1
return rows1 + rows2 as final

The above returns result if both has some records, if the node2 doesn't have any matching results then the final result is empty even though node1 have few results.

My requirement is to combine or union both which matches any of the condition. Could you please help me to achieve that.

Thanks in advance.

**Comment by burqen** [comment]
@selvarajendran I would love to help but this is sliding discussion is sidetracking away from topic a bit.
For specific help with queries can I recommend you seeking out
- The community site https://community.neo4j.com/
- Neo4j user slack channel https://neo4j-users-slack-invite.herokuapp.com/
Both of those spaces have lot's of people helping out with this type of questions :)

**Comment by ehx-v1** [comment]
Since fulltext indexes cover this, is there any point in keeping the issue open?

**Comment by InverseFalcon** [comment]
Fulltext schema indexes only allow indexes on string properties of relationships, so either this should remain open, or a new issue created for more general indexes on relationship properties regardless of the property type.

**Comment by alete89** [comment]
+1

**Comment by pirate** [comment]
This issue is getting a bit stale, but I just want to bump it and say there's still a desire for it.

Is there any hope for getting `UNIQUE CONSTRAINT` allowed on relationship properties?

For example in our model we have `(:USER)` nodes that can `[:TIP]`->`(:ARTIST)` nodes, and the TIP edges have a `{uuid}` property that we need to enforce at the database level. This is just one example, but actually all of our edges have `uuid`s (we use them to look up metadata related to that edge in other datastores), and we want to enforce uniqueness on them across the board without having to specify each src type, edge type, and dst type to build the constraints, and without having to manually `MERGE` relationships on every update. We also want to be able to do this: https://github.com/neo4j/neo4j/issues/11747

However this does not work:
```cypher
CREATE CONSTRAINT unique_edge_uuid IF NOT EXISTS ON ()-[edge]-() ASSERT edge.uuid IS UNIQUE
```

**Comment by burqen** [comment]
Hello everyone! Here's an update on this topic. We are far along with the work to have indexes on relationships. This is a necessary stepping stone to have uniqueness constraints on relationships although we haven't started work on that part yet.
I will keep this thread updated.
Thank you so much for your requests!

**Comment by burqen** [comment]
Good news everyone! As of Neo4j 4.3 you can index your relationships!

[Relationship Type Lookup Index](https://neo4j.com/docs/cypher-manual/current/administration/indexes-for-search-performance/#administration-indexes-create-a-relationship-type-lookup-index) will let you find all relationships of a given type with a single index scan. This will help immensely for queries where the type of the relationship creates the highest selectivity in your query, such as this:
```
MATCH (a)-[r:KNOWS]-(b) RETURN a, r, b
```

Relationship Property Indexes, on [single](https://neo4j.com/docs/cypher-manual/current/administration/indexes-for-search-performance/#administration-indexes-create-a-single-property-index-for-relationships) or [multiple](https://neo4j.com/docs/cypher-manual/current/administration/indexes-for-search-performance/#administration-indexes-create-a-composite-index-for-relationships) properties let's you find relationships with a given type and property value. It's based on a B+tree structure so it has great performance for predicates like `==`, `<`, `>`. This kind of index will help you a lot when you want to filter your result based on a relationship property such as this:
```
MATCH (a)-[r:KNOWS]-(b) WHERE r.since < 2011 RETURN a, r, b
```

You'll find more examples of how you can fine tune your queries using indexes [here](https://neo4j.com/docs/cypher-manual/current/query-tuning/indexes/).

It's still not possible to create uniqueness or node key constraints on relationships, but having indexes on relationships is a huge step towards making that possible in the future.

With this I will consider this issue solved 🎉 
Thank you so much for reporting everyone!
